serde/src/java/org/apache/hadoop/hive/serde2/objectinspector/ObjectInspectorConverters.java

```diff
@@ -349,32 +357,77 @@ public Object convert(Object input) {
    */
   public static class StructConverter implements Converter {
 
+    private static class StructFieldConverter implements Converter {
+
+      int inputIndex;
+      Converter fieldTypeConverter;
+
+      public StructFieldConverter(int inputIndex, Converter fieldTypeConverter) {
+        this.inputIndex = inputIndex;
+        this.fieldTypeConverter = fieldTypeConverter;
+      }
+
+      @Override
+      public Object convert(Object input) {
+        return fieldTypeConverter.convert(input);
+      }
+    }
+
     StructObjectInspector inputOI;
     SettableStructObjectInspector outputOI;
 
     List<? extends StructField> inputFields;
     List<? extends StructField> outputFields;
 
+    // if input and output schemas differ, resolve struct cols by name or by index
+    boolean resolveByName = false;
+
     ArrayList<Converter> fieldConverters;
 
     Object output;
 
     public StructConverter(ObjectInspector inputOI,
-        SettableStructObjectInspector outputOI) {
+        SettableStructObjectInspector outputOI, Configuration conf) {
       if (inputOI instanceof StructObjectInspector) {
+        if (conf != null) {
+          resolveByName = conf.getBoolean(HiveConf.ConfVars.HIVE_STRUCT_SCHEMA_CONVERSION_BY_NAME.varname, false);
+        }
+
         this.inputOI = (StructObjectInspector)inputOI;
         this.outputOI = outputOI;
         inputFields = this.inputOI.getAllStructFieldRefs();
         outputFields = outputOI.getAllStructFieldRefs();
 
-        // If the output has some extra fields, set them to NULL.
-        int minFields = Math.min(inputFields.size(), outputFields.size());
-        fieldConverters = new ArrayList<Converter>(minFields);
-        for (int f = 0; f < minFields; f++) {
-          fieldConverters.add(getConverter(inputFields.get(f)
-              .getFieldObjectInspector(), outputFields.get(f)
-              .getFieldObjectInspector()));
+        if (resolveByName) {
+          Map<String, Integer> inputFieldIndicesByName = new HashMap<>();
+          for (int f = 0; f < inputFields.size(); f++) {
+            inputFieldIndicesByName.put(inputFields.get(f).getFieldName().toLowerCase(), f);
+          }
+
+          fieldConverters = new ArrayList<>(outputFields.size());
+          for (int f = 0; f < outputFields.size(); f++) {
+            StructField out = outputFields.get(f);
```

Comment:
since index isn't used for something else than looping we can simplify the code
for (StructField out : outputFields) {